packages/gatsby/src/utils/babel-loader.js

```diff
@@ -48,11 +56,27 @@ module.exports = babelLoader.custom(babel => {
           ...options,
         },
       }
+
+      customOptionsCache.set(stage, toReturn)
+
+      return toReturn
     },
 
     // Passed Babel's 'PartialConfig' object.
     config(partialConfig, { customOptions }) {
+      let configCacheKey = customOptions.stage
+      if (partialConfig.hasFilesystemConfig()) {
+        // partialConfig.files is a Set that accumulates used config files (absolute paths)
+        partialConfig.files.forEach(configFilePath => {
+          configCacheKey += `_${configFilePath}`
+        })
+      }
```

Comment:
This is a bit speculative - I did check what `partialConfig.files` is, but I actually didn't test run this in scenario with `.babelrc` in nested directories (also main reason for making this a draft PR for now)

Reply:
Ok, this is not sufficent - in `develop` `.babelrc` filepaths might not change, but their content can (not sure if changing `.babelrc` file is "hot reloadable", but I would assume it is (?) )

Reply:
Added memoization cache invalidation in https://github.com/gatsbyjs/gatsby/pull/28738/commits/eebbbfac88106489a70dea8eedd20730439bd1de